cnap/core/inferqueue.py

```diff
@@ -0,0 +1,374 @@
+                LOG.info("Reconnect to Redis server")
+
+                # Sleep before reconnect, double the sleep time per reconnection.
+                sleep_time *= 2
+
+                time.sleep(sleep_time)
+
+                # Connect to Redis server.
+                self._conn = redis.Redis(host=host, port=port, db=0)
+
+                # Connection is not successful, can not return.
+                continue
+
+            # If no connection error occurs, return directly.
+            return
+
+
+    def publish_frame(self, topic: str, frame: Frame) -> None:
+        if topic is None:
+            raise ValueError("topic can not be None")
+        if frame is None:
+            raise ValueError("frame can not be None")
+        try:
+            frame_blob = frame.to_blob()
+        except RuntimeError as e:
+            raise RuntimeError(e) from e
+        self._conn.rpush(topic, frame_blob)
+
+    def get_frame(self, topic: str) -> Frame:
+        if topic is None:
+            raise ValueError("topic can not be None")
+        msg = self._conn.lpop(topic)
+        if msg is None:
+            return None
+        try:
+            return Frame.from_blob(msg)
+        except TypeError as e:
+            raise TypeError(e) from e
+        except RuntimeError as e:
+            raise RuntimeError(e) from e
+
+    def drop(self, topic: str) -> int:
+        if topic is None:
+            raise ValueError("topic can not be None")
+        queue_len = self._conn.llen(topic)
+        drop_frame = 0
+        if queue_len > self.buffer_len:
+            self._conn.ltrim(topic, 0 - self.buffer_len, -1)
+            drop_frame = queue_len - self.buffer_len
+        return drop_frame
+
+    def infer_queue_available(self, topic: str) -> bool:
+        if topic is None:
+            raise ValueError("topic can not be None")
+        key = topic + "-available"
+        return self._conn.exists(key)
+
+    def register_infer_queue(self, topic: str):
+        if topic is None:
+            raise ValueError("topic can not be None")
+        key = topic + "-available"
+        if self._conn.exists(key):
+            count = int(self._conn.get(key)) + 1
+            self._conn.set(key, count)
+        else:
+            LOG.info("Register inference queue for topic: %s", topic)
+            self._conn.set(key, 1)
+
+    def unregister_infer_queue(self, topic: str):
+        if topic is None:
+            raise ValueError("topic can not be None")
+        key = topic + "-available"
+        count = int(self._conn.get(key)) - 1
+        if count > 0:
+            self._conn.set(key, count)
+        else:
+            LOG.info("Unregister inference queue for topic: %s", topic)
+            self._conn.delete(key)
+            self._conn.delete(topic)
+
+class KafkaInferQueueClient(InferQueueClientBase):
+    """
+    Kafka implementation for inference queue client.
+    """
+
+    MAX_RECONNECTION_TIMES = 5
```

Comment:
Duplicate definition, may be we declare it in base class?

Reply:
Thanks, fixed.